quic/s2n-quic-core/src/buffer/receive_buffer.rs

```diff
@@ -0,0 +1,367 @@
+
+            buffer.split_to(watermark)
+        })
+    }
+
+    /// Pops a buffer from the front of the receive queue as long as the `transform` function returns a
+    /// non-empty buffer.
+    #[inline]
+    fn pop_transform<F: Fn(&mut BytesMut) -> BytesMut>(
+        &mut self,
+        transform: F,
+    ) -> Option<BytesMut> {
+        let slot = self.slots.front_mut()?;
+
+        if !slot.is_occupied(self.start_offset) {
+            return None;
+        }
+
+        let buffer = slot.data_mut();
+
+        let out = transform(buffer);
+
+        slot.add_start(out.len());
+
+        if slot.start() == slot.end_allocated() {
+            // remove empty buffers
+            self.slots.pop_front();
+        }
+
+        // filter out empty buffers
+        if out.is_empty() {
+            return None;
+        }
+
+        self.start_offset += out.len() as u64;
+
+        Some(out)
+    }
+
+    /// Returns the amount of data that had already been consumed from the
+    /// receive buffer.
+    #[inline]
+    pub fn consumed_len(&self) -> u64 {
+        self.start_offset
+    }
+
+    /// Returns the total amount of consecutive received data.
+    /// This includes the already consumed data as well as the data that is still
+    /// buffered and available for consumption.
+    #[inline]
+    pub fn total_received_len(&self) -> u64 {
+        self.consumed_len() + self.len() as u64
+    }
+
+    /// Resets the receive buffer.
+    /// This will drop all previously received data.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.slots.clear();
+        self.start_offset = Default::default();
+    }
+
+    #[inline(always)]
+    fn insert(&mut self, idx: usize, slot: Slot) {
+        if self.slots.len() < idx {
+            debug_assert_eq!(self.slots.len() + 1, idx);
+            self.slots.push_back(slot);
+        } else {
+            self.slots.insert(idx, slot);
+        }
+    }
+
+    #[inline]
+    fn allocate_request(&mut self, mut idx: usize, mut request: Request) {
+        while !request.is_empty() {
+            let size = Self::allocation_size(request.start());
+            let offset = Self::align_offset(request.start(), size);
+            let buffer = BytesMut::with_capacity(size);
+            let end = offset + size as u64;
+            let mut slot = Slot::new(offset, end, buffer);
+
+            let slot::Outcome { lower, mid, upper } = slot.try_write(request);
+
+            debug_assert!(lower.is_empty());
+
+            self.insert(idx, slot);
+            idx += 1;
+
+            if let Some(mid) = mid {
+                self.insert(idx, mid);
+                idx += 1;
+            }
+
+            request = upper;
+        }
+    }
+
+    /// Aligns an offset to a certain alignment size
+    #[inline]
+    fn align_offset(offset: u64, alignment: usize) -> u64 {
+        (offset / (alignment as u64)) * (alignment as u64)
+    }
+
+    #[inline]
+    fn allocation_size(offset: u64) -> usize {
```

Comment:
Could you add some of the text from the PR (ie `..gradually increases the allocation sizes as the offset increases. This is under the assumption that streams that receive a lot of data will continue to receive a lot of data`) into a comment on this function?

Reply:
Yep good call